### DISCUSSION TITLE
Project Panama

### DISCUSSION THREAD

**Discussion by javster101** [LWJGL/lwjgl3#609] [General]
Is there a plan to migrate the existing native accessto  Project Panama with JDK 16?

**Comment by octylFractal** [comment]
Yes, not with JDK 16 as they're still incubating modules, but you can search the LWJGL forums and the issue tracker here with the keyword "panama" to see some of the rough ideas.

**Comment by apostolos** [comment]
Quoting @Spasi 

> Project Panama consists of two projects: the Foreign-Memory Access API (https://openjdk.java.net/jeps/370) and the Vector API (https://openjdk.java.net/jeps/338). 370 has been incubating since JDK 14, 338 is being introduced in JDK 16 (also incubating). I've been tracking the progress of both projects for several years and the plan is to migrate LWJGL to 370 when the API stabilizes and we're closer to a non-incubating release. It's not going to be LWJGL 3 anymore though, since we're talking about a massive breaking change (e.g. all NIO buffers are being replaced with 370's MemorySegment type). For a while, LWJGL 3 (current API) and LWJGL 4 (Panama-based API) will be maintained in parallel.
> 
> I'm not in a rush to start LWJGL 4 though, for a few reasons:
> 
> - The Panama API is still a work-in-progress, it changes quite often.
> - Performance is not quite there yet. There are several reasons for that, but they're actively working on it on the JVM side. I'm looking for at least performance parity with JNI.
> - Panama is designed to have strong synergy with Project Valhalla (https://openjdk.java.net/projects/valhalla/), i.e. value types. Valhalla is another major project under development for several years, but that one is extremely complicated, with major technical difficulties. Ideally, when LWJGL 4 is released, we'll have both Panama and Valhalla available.

**Reply by Glavo** [reply]
How about now? Panama has now been moved to the `java.base` module. Although it is still in the preview state, it is easier to use than the incubator state.

At present, Panama is still not perfect, it needs more feedback from real world use cases. If we can start to explore LWJGL4 before Panama is officially finalized, I believe it can help Panama find some design defects, so that the design of Panama can be closer to perfection before the API is frozen.

**Reply by Spasi** [reply]
Hey @Glavo,

See [this reply](https://github.com/LWJGL/lwjgl3/issues/853#issuecomment-1431660012). The first item is done already, now working on getting everything ready for the 3.3.2 release.

**Reply by Glavo** [reply]
Latest progress:

https://mail.openjdk.org/pipermail/panama-dev/2023-July/019510.html

Foreign Function & Memory API will  exit preview in Java 22.

**Reply by theoparis** [reply]
Java 22 is out now 👀

**Reply by Glavo** [reply]
Latest progress: Value Objects ([JEP 401](https://openjdk.org/jeps/401)) will begin previewing in JDK 28.

**Comment by dustContributor** [comment]
Does anybody knows if the new interfaces provided in JEP 370 and in JEP 338 are going to be switched to value types once they come out?

They look like it since they all mention being "value based", like the "new" time API.

**Reply by octylFractal** [reply]
They should be, see https://openjdk.java.net/jeps/390 which warns on any `@ValueBased` class.

**Reply by dustContributor** [reply]
I see. Yeah it'd make total sense to do it. Same with the vectorization
API. As is currently it depends in too much JIT magic to avoid silly heap
allocations.

**Reply by ws909** [reply]
Just for reference, [JEP 438](https://openjdk.org/jeps/438) (5th Vector incubator) specifically states that the Vector API is now only being held back by the lack of Valhalla support:

> This JEP proposes to re-incubate the API in JDK 20, with no changes in the API relative to JDK 19. The implementation includes small set of bug fixes and performance enhancements. This JEP also clarifies that alignment with Project Valhalla is a critical part of completing the Vector API.

**Comment by Eclesia** [comment]
Hello, I am coying here a message I wrote here : https://github.com/club-doki7/vulkan4j/discussions/177

I am also interested in this LWJGL <> Vulkan4j discussion.
The panama API is great and I made several lib bindings using it.

One that may interest you is the one for [Khronos ANARI-SDK](https://github.com/KhronosGroup/ANARI-SDK)
The java panama binding is here : https://bitbucket.org/Eclesia/anari-java/src/main/
I checked vulkan4j code and it's coding style and classes for ffm are very close to what I have, it would make a nice addition.

But I have some concerns, as a user I would love this code to be in LWJGL 4 because there is a huge community and many projects using it (JME and more).

Would it be possible to merge the 3 projects (Vulkan4j,LWJGL4,AnariJ) ?
to create something new and neutral ?

I'm asking because I've been doing some 3D rendering since JOGL 1 (15years ago ?) and seen many 3d binding libs raise and die. (Java 3D, JavaFX 3D, Ardor3D, Libgdx, ...and more...)

Vulkan4j is nice, but doesn't have a community. LWJGL panama is still empty, but has a large community.
Have you try to talk with them ?

Isn't it time to start LWJGL 4 (or any new name you want to give it) ?
